### PR TITLE
csi/api: populate ReadAllocs/WriteAllocs fields

### PR DESCRIPTION
**Pull request by tgross**
Fixes https://github.com/hashicorp/nomad/issues/9215

The API is missing values for `ReadAllocs` and `WriteAllocs` fields, resulting
in allocation claims not being populated in the web UI. These fields mirror
the fields in `nomad/structs.CSIVolume`. Returning a separate list of stubs
for read and write would be ideal, but this can't be done without either
bloating the API response with repeated full `Allocation` data, or causing a
panic in previous versions of the CLI.

The `nomad/structs` fields are persisted with nil values and are populated
during RPC, so we'll do the same in the HTTP API and populate the `ReadAllocs`
and `WriteAllocs` fields with a map of allocation IDs, but with null
values. The web UI will then create its `ReadAllocations` and
`WriteAllocations` fields by mapping from those IDs to the values in
`Allocations`, instead of flattening the map into a list.

---

I've tested this out via the web UI and verified `nomad volume status`, `nomad alloc status`, and `nomad node status` aren't broken, including with a v0.12.8 version of the command line.

### CODE REVIEW

File: ui/app/serializers/volume.js [hunk 1]
```diff
@@ -23,17 +23,16 @@ export default class VolumeSerializer extends ApplicationSerializer {
...
-      alloc.ID = id;
+    const bindIDToAlloc = allocList => id => {
+      const alloc = allocList.find(a => a.ID === id);
       return alloc;
     };
 
-    hash.ReadAllocations = Object.keys(readAllocs).map(bindIDToAlloc(readAllocs));
-    hash.WriteAllocations = Object.keys(writeAllocs).map(bindIDToAlloc(writeAllocs));
+    hash.ReadAllocations = Object.keys(readAllocs).map(bindIDToAlloc(hash.Allocations));
+    hash.WriteAllocations = Object.keys(writeAllocs).map(bindIDToAlloc(hash.Allocations));
```
Comment (tgross) [inline ui/app/serializers/volume.js:35]: Question for reviewers: would it save browser memory if we nulled-out the `hash.Allocations`, `hash.ReadAllocs` and `hash.WriteAllocs` fields after we're done with them here?

File: api/csi.go [hunk 2]
```diff
@@ -110,11 +107,17 @@ type CSIVolume struct {
...
-	ReadAllocs  map[string]*Allocation
+	// ReadAllocs is a map of allocation IDs for tracking reader claim status.
+	// The Allocation value will always be nil; clients can populate this data
+	// by iterating over the Allocations field.
+	ReadAllocs map[string]*Allocation
+
+	// WriteAllocs is a map of allocation IDs for tracking writer claim
+	// status. The Allocation value will always be nil; clients can populate
+	// this data by iterating over the Allocations field.
 	WriteAllocs map[string]*Allocation
```
Comment (notnoop) [inline api/csi.go:118]: Curious why we chose this representation?  If the map value is never a populated alloc, how does it compare to using `map[string]bool` (explicitly meaningless value) or a plain slice `[]string` (avoid value but losing uniqueness)? Not advocating for changing it, specially if we are constrained by compatibility.

**Review by notnoop — APPROVED** [review]
The golang side is LGTM.

File: api/csi.go (hunk 2 above)
Comment (tgross) [reply api/csi.go:118]: It's solely constrained by compatibility at this point, unfortunately.

With 20/20 hindsight, we would have had the `nomad/structs` version store only the `CSIVolumeClaim`s and not allocations (which has been a huge complication with respect to garbage collected jobs and allocations), and then had this API struct return `WriteAllocs map[string]*AllocationListStub` and `ReadAllocs map[string]*AllocationListStub`, without an `Allocations` field at all. See https://github.com/hashicorp/nomad/issues/8734 for more

File: ui/app/serializers/volume.js [hunk 3]
```diff
@@ -23,17 +23,16 @@ export default class VolumeSerializer extends ApplicationSerializer {
...
 
-    // Convert hash-based allocation embeds to lists
+    // Populate read/write allocation lists from aggregate allocation list
     const readAllocs = hash.ReadAllocs || {};
     const writeAllocs = hash.WriteAllocs || {};
-    const bindIDToAlloc = hash => id => {
-      const alloc = hash[id];
-      alloc.ID = id;
+    const bindIDToAlloc = allocList => id => {
+      const alloc = allocList.find(a => a.ID === id);
```
Comment (DingoEatingFuzz) [inline ui/app/serializers/volume.js:30]: I'm not sure if the allocs list will ever be large enough for this to be significant, but using a `find` here changes the complexity from linear to quadratic in the worst case. 

It should be possible to invert this and iterate over allocs and build the `ReadAllocations` and `WriteAllocations` representations by checking for existence in the `ReadAllocs` and `WriteAllocs` hashes. That would make this linear again.

File: ui/app/serializers/volume.js (hunk 1 above)
Comment (DingoEatingFuzz) [reply ui/app/serializers/volume.js:35]: I don't know off the top of my head if hash sticks around or if a copy is what gets inserted into the store. It wouldn't hurt to add a `delete hash.Allocations` just to err on the side of caution.

**Review by DingoEatingFuzz — APPROVED** [review]
I had a note on performance, but I'll let you decide if it's worth changing.

File: ui/app/serializers/volume.js (hunk 3 above)
Comment (tgross) [reply ui/app/serializers/volume.js]: Fixed in 198264e

**Pull request merged**